### PR TITLE
Deprecate the special SYMFONY__ environment variables

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #20089 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Just challenging this one: the pending PR on this topic now says:
> You can also define the default value of any existing parameters using special environment variables named after their corresponding parameter prefixed with SYMFONY__ after replacing dots by double underscores (e.g. SYMFONY__KERNEL__CHARSET to set the default value of the kernel.charset parameter). These default values are resolved when compiling the service container and won't change at runtime once dumped.

See http://pr-6918-6qmocelev2lwe.eu.platform.sh/configuration/external_parameters.html#environment-variables

To me, the use case is different. `env()` parameters do not allow this (setting the default value of parameter at compile time through env vars).
Do we really consider this as something we don't want to support anymore?

**Comment by nicolas-grekas** [comment]
If the answer is yes (see above challenge), shouldn't we deprecate `EnvParametersResource` and `getEnvParameters` also?

**Comment by iltar** [comment]
@nicolas-grekas if `SYMFONY__` indicated that it was a value only used to build parameters and stores the calculated version, maybe something like `build_env(...)` could solve the problem?

**Comment by nicolas-grekas** [comment]
It's also a matter of who own the relationship between the env and the config.
`env()` or any such alternatives are the way to go to *pull* from the env.
`SYMFONY__` *push* to parameters.
Different directions, different use case to me.

**Comment by ro0NL** [comment]
> It's also a matter of who owns the relationship between the env and the config.

Deprecating the magic `SYMFONY__` vars brings back the ownership to the config.. imo that's a good thing.

**Comment by wouterj** [comment]
If deprecation is coming, I don't agree with the implementation of the deprecation. I can imagine people want to keep the `SYMFONY__` parameters and just instead request them with `env('SYMFONY__')` (to ease migration). This will currently still result in the deprecation...

**Comment by nicolas-grekas** [comment]
I propose #20618 as an alternative to using `SYMFONY__` env vars for compile time env vars resolution. Which means I'm not :+1: to deprecate them, thus unlocking this PR from my pov.

**Comment by nicolas-grekas** [comment]
@wouterj can you think of another way? IMO, it looks sensible the way it is now.

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 1]
```diff
@@ -537,6 +537,7 @@ protected function getEnvParameters()
         $parameters = array();
         foreach ($_SERVER as $key => $value) {
             if (0 === strpos($key, 'SYMFONY__')) {
+                @trigger_error(sprintf('The special environment variables that start with SYMFONY__ (such as "%s") are deprecated as of 3.2 and will be removed in 4.0. Use instead the %env()% syntax to get the value of environment variables in configuration files.', $key), E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/Kernel.php]: 3.3

File: src/Symfony/Component/HttpKernel/Tests/KernelTest.php [hunk 2]
```diff
@@ -718,6 +719,25 @@ public function testTerminateDelegatesTerminationOnlyForTerminableInterface()
...
+        $_SERVER['SYMFONY__FOO__BAR'] = 'baz2';
+
+        $kernel = $this->getKernel();
+        $method = new \ReflectionMethod($kernel, 'getEnvParameters');
+        $method->setAccessible(true);
+
+        $deprecationMessage = 'SYMFONY__ special environment variables are deprecated as of 3.2 and will be removed in 4.0. Use instead the %env()% syntax to get the value of environment variables in configuration files.';
+        $expectedDeprecations = array($deprecationMessage, $deprecationMessage);
+
+        ErrorAssert::assertDeprecationsAreTriggered($expectedDeprecations, function () use ($kernel, $method) {
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/Tests/KernelTest.php]: You now have to use the annotation instead.

**Review by xabbuh — CHANGES_REQUESTED** [review]

**Comment by nicolas-grekas** [comment]
don't forget this one @javiereguiluz

**Comment by javiereguiluz** [comment]
@nicolas-grekas updated with the fixes suggested by reviewers.

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 3]
```diff
@@ -537,6 +537,7 @@ protected function getEnvParameters()
         $parameters = array();
         foreach ($_SERVER as $key => $value) {
             if (0 === strpos($key, 'SYMFONY__')) {
+                @trigger_error(sprintf('The special environment variables that start with SYMFONY__ (such as "%s") are deprecated as of 3.3 and will be removed in 4.0. Use instead the %env()% syntax to get the value of environment variables in configuration files.', $key), E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/Kernel.php]: I would move "instead" after "syntax".

File: src/Symfony/Component/HttpKernel/Tests/KernelTest.php [hunk 4]
```diff
@@ -719,6 +719,22 @@ public function testTerminateDelegatesTerminationOnlyForTerminableInterface()
     }
 
     /**
+     * @group legacy
+     * @expectedExceptionMessage The special environment variables that start with SYMFONY__ (such as "SYMFONY__FOO__BAR") are deprecated as of 3.3 and will be removed in 4.0. Use instead the %env()% syntax to get the value of environment variables in configuration files.
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/Tests/KernelTest.php]: `@expectedDeprecation`

**Review by xabbuh — CHANGES_REQUESTED** [review]

**Comment by ro0NL** [comment]
> If deprecation is coming, I don't agree with the implementation of the deprecation. I can imagine people want to keep the SYMFONY__ parameters and just instead request them with env('SYMFONY__') (to ease migration). This will currently still result in the deprecation...

This is true. Defining an `SYMFONY__` env var is not deprecated, only the magic part where it's set as a parameter.

I also dont really agree on tests here.. using reflection smells.

I think roughly the path is
- Remove `$_SERVER` implementation
- Trigger deprecation if a `SYMFONY__` var is defined, but not configured as a parameter, ie;

```yaml
'database.name': '...'
```

- Otherwise set de default parameter along with it, ie;

```yaml
'database.name': '%env(SYMFONY__DATABASE_NAME)%'
```

Should work out right?

**Comment by nicolas-grekas** [comment]
Let be pragmatic here: if any `SYMFONY__` env var exists, its because *we* asked for them. Which means it's OK to deprecate *the env vars*. MHO.

**Comment by ro0NL** [comment]
:+1: ok then. Already thought something like that.. just wanted to make sure we did not overlooked it :)

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 3 above)
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/Kernel.php]: ` and will be removed in 4.0`

We cant really remove user env vars ^^ i'd go with `and will not be supported anymore in 4.0`.

**Comment by chalasr** [comment]
Instead of triggering the deprecation when registering the parameter, couldn't we trigger it when getting the parameter (reversing the key transformation and trigger if we find it in $_SERVER)?

Also, shouldn't the `Kernel::getEnvParameters()` method be deprecated?

**Comment by nicolas-grekas** [comment]
rebase needed - to make tests green at least

**Comment by javiereguiluz** [comment]
The rebase wen't wrong ... closing it in favor of #21889.

**Pull request closed without merging**